radix-transaction-scenarios/src/executor.rs

```diff
@@ -232,169 +185,177 @@ where
         self
     }
 
-    /// A callback that is called when a new scenario is started.
-    pub fn on_scenario_start<F: FnMut(&ScenarioMetadata) + 'a>(mut self, callback: F) -> Self {
-        self.on_scenario_start = Box::new(callback);
+    /// A callback that is called when a scenario is started.
+    pub fn on_scenario_started<F: FnMut(&ScenarioMetadata) + 'a>(mut self, callback: F) -> Self {
+        self.on_scenario_started = Box::new(callback);
         self
     }
 
-    pub fn without_protocol_updates(self) -> Self {
-        self.on_new_protocol_requirement_encountered(|_, _, _| {})
+    /// A callback that is called when a scenario is ended.
+    pub fn on_scenario_ended<F: FnMut(&ScenarioMetadata, &EndState, &D) + 'a>(
+        mut self,
+        callback: F,
+    ) -> Self {
+        self.on_scenario_ended = Box::new(callback);
+        self
     }
 
-    /// A callback that is called after bootstrapping if bootstrap is enabled.
-    pub fn after_bootstrap<F: FnMut(&NetworkDefinition, &mut D) + 'a>(
+    /// A callback that is called before a protocol update is executed.
+    pub fn on_before_protocol_update_executed<F: FnMut(&ProtocolUpdateExecutor) + 'a>(
         mut self,
         callback: F,
     ) -> Self {
-        self.after_bootstrap = Box::new(callback);
+        self.on_before_protocol_update_executed = Box::new(callback);
         self
     }
 
-    pub fn execute_all_matching(
-        self,
-        filter: ScenarioFilter,
-    ) -> Result<ScenarioExecutionReceipt<D>, ScenarioExecutorError> {
-        self.internal_execute(Some(filter))
+    pub fn into_database(self) -> D {
+        self.database
     }
 
-    pub fn execute_all(self) -> Result<ScenarioExecutionReceipt<D>, ScenarioExecutorError> {
-        self.internal_execute(None)
+    /// Each scenario is executed once, when it first becomes valid.
+    pub fn execute_every_protocol_update_and_scenario(
+        &mut self,
+    ) -> Result<(), ScenarioExecutorError> {
+        self.execute_protocol_updates_and_scenarios(
+            ProtocolBuilder::for_network(&self.network_definition).until_latest_protocol_version(),
+            ScenarioTrigger::AtStartOfEveryProtocolVersion,
+            ScenarioFilter::AllScenariosFirstValidAtProtocolVersion,
+        )
     }
 
-    fn internal_execute(
-        mut self,
-        filter: Option<ScenarioFilter>,
-    ) -> Result<ScenarioExecutionReceipt<D>, ScenarioExecutorError> {
-        // Bootstrapping if needed
-        if self.bootstrap {
-            Bootstrapper::new(
-                self.network_definition.clone(),
-                &mut self.database,
-                VmInit::new(&self.scrypto_vm, self.native_vm_extension.clone()),
+    pub fn execute_protocol_updates_and_scenarios(
+        &mut self,
+        protocol_executor: ProtocolExecutor,
+        trigger: ScenarioTrigger,
+        filter: ScenarioFilter,
+    ) -> Result<(), ScenarioExecutorError> {
+        Bootstrapper::new(
```

Comment:
Well the same applies to protocol updates. The correct solution would be to pass a “start at protocol version” parameter (not simply “bootstrap yes/no”). But we have no use for it at the moment in any case :)

As per the PR description, I want to move bootstrapping into the protocol builder anyway, so this would be a natural place for it.

Or we could eventually be storing the current protocol version (slash current  upgrade status if partway through an update) in the substate store which would allow for resumption.

Reply:
I agree with you that the more ideal solution would be being able to start at a particular protocol version, but my primary concern here is that in the current implementation running scenarios on top of an already-bootstrapped database does not work and the execution fails with an error and I do not think that we should make the assumption that the scenario executor will always be bootstrapping the database.

Perhaps a simpler solution until we make genesis/babylon a protocol update is to ignore this bootstrapping error if it does occur.